Logstash/logstash-input-azurewadtable/lib/logstash/inputs/azurewadtable.rb

```diff
@@ -63,99 +75,103 @@ def build_latent_query
       query_filter << " or (PartitionKey gt '#{i.to_s.rjust(19, '0')}___#{partitionkey_from_datetime(@last_timestamp)}' and PartitionKey lt '#{i.to_s.rjust(19, '0')}___#{partitionkey_from_datetime(@until_timestamp)}')"
     end # for block
     query_filter = query_filter.gsub('"','')
-    query_filter
-  end
-
-  def build_zero_latency_query
-    @logger.debug("from #{@last_timestamp} to most recent data")
-    # query data using start_from_time
-    query_filter = "(PartitionKey gt '#{partitionkey_from_datetime(@last_timestamp)}')"
-    for i in 0..99
-      query_filter << " or (PartitionKey gt '#{i.to_s.rjust(19, '0')}___#{partitionkey_from_datetime(@last_timestamp)}' and PartitionKey lt '#{i.to_s.rjust(19, '0')}___9999999999999999999')"
-    end # for block
-    query_filter = query_filter.gsub('"','')
-    query_filter
+    return AzureQuery.new(@logger, @azure_table_service, @table_name, query_filter, @last_timestamp.to_s + "-" + @until_timestamp.to_s, @entity_count_to_process)
   end
 
   def process(output_queue)
-    if @data_latency_minutes > 0
-      @until_timestamp = (Time.now - (60 * @data_latency_minutes)).iso8601 unless @continuation_token
-      query_filter = build_latent_query
-    else
-      query_filter = build_zero_latency_query
+    @until_timestamp = (Time.now - (60 * @data_latency_minutes)).iso8601
+    last_good_timestamp = nil
+
+     # split first query so we don't fetch old data several times for no reason
+    if @first_run
+      diff = DateTime.iso8601(@until_timestamp).to_time - DateTime.iso8601(@last_timestamp).to_time
+      if diff > INITIAl_QUERY_SPLIT_PERIOD_MINUTES * 60
+        @logger.debug("Splitting initial query in two")
+        original_until = @until_timestamp
+        
+        @until_timestamp = (DateTime.iso8601(@until_timestamp).to_time - INITIAl_QUERY_SPLIT_PERIOD_MINUTES * 60).iso8601
+
+        query = build_latent_query
+        @duplicate_detector.filter_duplicates(query, ->(entity) {
+          on_new_data(entity, output_queue, last_good_timestamp)
+        }, false)
+
```

Comment:
I still not fully sure about the point of the this first split query, here is what I am guessing:
We use this query to catch all partitions that occurred in the storage account in the past, and has no chance of new data coming in. This query isn't added to the cache and so it isn't rerun. INITIAL_QUERY_SPLIT_PERIOD_MINUTES is 30 minutes just means, after 30 min, we are highly confident no more data should come into these partitions?

Reply:
yes, you are right - the 30 minutes cut off point is there because we are highly confident there's no data that will arrive in those partitions.
The first query is split because the user might require data from a point very far in the past and it doesn't really make sense to run that expensive query five times, knowing that the odds of finding new events are ~ 0